lib/services/sessiontracker.go

```diff
@@ -44,3 +50,118 @@ type SessionTrackerService interface {
 	// UpdatePresence updates the presence status of a user in a session.
 	UpdatePresence(ctx context.Context, sessionID, user string) error
 }
+
+// UpdateSessionTrackerState is a helper function to add a session tracker participant.
+func AddSessionTrackerParticipant(ctx context.Context, sts SessionTrackerService, sid string, participant *types.Participant) error {
+	req := &proto.UpdateSessionTrackerRequest{
+		SessionID: sid,
+		Update: &proto.UpdateSessionTrackerRequest_AddParticipant{
+			AddParticipant: &proto.SessionTrackerAddParticipant{
+				Participant: participant,
+			},
+		},
+	}
+
+	err := sts.UpdateSessionTracker(ctx, req)
+	return trace.Wrap(err)
+}
+
+// UpdateSessionTrackerState is a helper function to remove a session tracker participant.
+func RemoveSessionTrackerParticipant(ctx context.Context, sts SessionTrackerService, sid string, participantID string) error {
+	req := &proto.UpdateSessionTrackerRequest{
+		SessionID: sid,
+		Update: &proto.UpdateSessionTrackerRequest_RemoveParticipant{
+			RemoveParticipant: &proto.SessionTrackerRemoveParticipant{
+				ParticipantID: participantID,
+			},
+		},
+	}
+
+	err := sts.UpdateSessionTracker(ctx, req)
+	return trace.Wrap(err)
+}
+
+// UpdateSessionTrackerState is a helper function to update a session tracker state.
+func UpdateSessionTrackerState(ctx context.Context, sts SessionTrackerService, sid string, state types.SessionState) error {
+	req := &proto.UpdateSessionTrackerRequest{
+		SessionID: sid,
+		Update: &proto.UpdateSessionTrackerRequest_UpdateState{
+			UpdateState: &proto.SessionTrackerUpdateState{
+				State: state,
+			},
+		},
+	}
+
+	err := sts.UpdateSessionTracker(ctx, req)
+	return trace.Wrap(err)
+}
+
+// UpdateSessionTrackerExpiry is a helper function to update a session tracker expiry.
+func UpdateSessionTrackerExpiry(ctx context.Context, sts SessionTrackerService, sid string, expires time.Time) error {
+	req := &proto.UpdateSessionTrackerRequest{
+		SessionID: sid,
+		Update: &proto.UpdateSessionTrackerRequest_UpdateExpiry{
+			UpdateExpiry: &proto.SessionTrackerUpdateExpiry{
+				Expires: &expires,
+			},
+		},
+	}
+
+	err := sts.UpdateSessionTracker(ctx, req)
+	return trace.Wrap(err)
+}
+
+// TrackSession creates the given session tracker. The session tracker's expiration
+// will be extended on each tick from the given ticker until a termiante signal is
+// received. Then the tracker's state will be updated to terminated.
+func TrackSession(ctx context.Context, sts SessionTrackerService, tracker types.SessionTracker, ticker clockwork.Ticker, terminate <-chan struct{}) error {
```

Comment:
Yeah I did a bit too much "experimental" coding with this PR, thanks for your patience with the review.

I thought this would be simpler to export to a reusable function, but it ended up needing all three of these primitives (if ctx is canceled then `UpdateSessionTracker` doesn't have a valid ctx to use). I'll see if there's a simpler way but I'll probably just tear it back down.

Reply:
No worries.

A totally separate context could be OK, with a bit more bookkeeping to cancel it if the original one gets cancelled. Let me know, though, it's easy to make up things without coding them. :)

Reply:
This is now handled by `UpdateExpirationLoop(ctx context.Context, clock clockwork.Clock)`, which is much more straightforward.